### PR TITLE
Add support for Stream Deck Mini

### LINKED ISSUE #2100: Stream Deck Mini HID values
The Stream Deck mini has a different HID identifiers than the regular:

```
                Stream Deck Mini:

                  Product ID: 0x0063
                  Vendor ID: 0x0fd9  (Elgato Systems GmbH)
                  Version: 1.00
                  Serial Number: BL36H1A26790
                  Speed: Up to 480 Mb/sec
                  Manufacturer: Elgato Systems
                  Location ID: 0x14123000 / 18
                  Current Available (mA): 500
                  Current Required (mA): 300
                  Extra Operating Current (mA): 0
```

### PR DESCRIPTION
**Pull request by octplane**
This is a follow-up of #2100

### CODE REVIEW

**Comment by octplane** [comment]
More details about the difference in implementation: https://github.com/Lange/node-elgato-stream-deck/issues/56#issuecomment-412263336

**Comment by octplane** [comment]
The vertical offset seems to be due to some bytes of additional data at image start, as if we were sending too many bytes. I'll try removing the BMP header, to see if it changes anything(?).

**Comment by octplane** [comment]
Of course, NSImage+BMP already rotates the image in some kind 🌠

**Comment by octplane** [comment]
I have a first version that works correctly with my device. I'm now refactoring the code to support the regular and mini devices. 
IMG_1912

**Comment by octplane** [comment]
This is now ready for review. Every feature seems to work correctly on my Stream Deck Mini and hopefully I didn't break anything for other users.

Few things that could/should be added:
- validate that the button is in the valid device range. I can implement that if needed.
- export the `isMini` field to Lua as I have no idea how to do it
- the way the `buttonNumber` is computed is not really nice. Other implementer seem to imply that the value received can be split in some way but I really have no idea, esp. as the HID intricacies are far beyond my graps...

**Comment by octplane** [comment]
This is now ready for review. Every feature seems to work correctly on my Stream Deck Mini and hopefully I didn't break anything for other users.

Few things that could/should be added:
- ensure that the user uses a supported button. I can implement that if needed.
- export the `isMini` field to Lua as I have no idea how to do it
- the way the `buttonNumber` is computed is not really nice. Other implementer seem to imply that the value received can be split in some way but I really have no idea, esp. as the HID intricacies are far beyond my graps...

**Comment by octplane** [comment]
- Image drawing performance is not as good as with the native application (I can clearly see the redraw happening).
- I have rebuilt in Release mode, and this is still visible. No idea why :(

**Comment by latenitefilms** [comment]
I currently only own the original Stream Deck, however I'll try get my hands on a Stream Deck Mini and test.

**Comment by octplane** [comment]
- I traced the HID packets and image transfert seems to run usually as fast as the native app. But the redraws seems more visible
- I have fixed an indexing difference between the mini and regular device
- There is still an hidden bug that prevents the picture from showing completely: some pixels are missing in the bottom left corner (corresponds to the the actual bottom right in the source image)

**Comment by octplane** [comment]
- pictures are now displayed correctly
- the performance issue is still visible
- also, I can see the Stream Deck bliting the previous drawn image in the new key image before overwriting it step by step. This is pretty puzzling.
- I shot high speed movies of the redraws, the original app displays 6 icons in 1/2 hi-speed frames whereas this branch's code takes several 10s of frames...

**Comment by latenitefilms** [comment]
Has the performance issue got something to do with the fact the original code [sent the images in two halves](https://github.com/Hammerspoon/hammerspoon/pull/2102/files#diff-aa78f02549f5b46cdaf02e368a6a834dL161)?

@cmsj might have some ideas on this?

File: extensions/streamdeck/HSStreamDeckDevice.m [hunk 1]
```diff
@@ -48,6 +54,42 @@ - (void)deviceDidSendInput:(NSNumber*)button isDown:(NSNumber*)isDown {
     _lua_stackguard_exit(skin.L);
 }
 
+- (int)targetSize {
+    if(self.isMini) {
+        return 80;
+    }
+    return 77;
```
Comment (latenitefilms) [inline extensions/streamdeck/HSStreamDeckDevice.m]: Should this be 72?

Comment (octplane) [reply extensions/streamdeck/HSStreamDeckDevice.m]: 👍

**Comment by octplane** [comment]
> Has the performance issue got something to do with the fact the original code [sent the images in two halves](https://github.com/Hammerspoon/hammerspoon/pull/2102/files#diff-aa78f02549f5b46cdaf02e368a6a834dL161)?
> 
> @cmsj might have some ideas on this?

This change is one of the major change between both devices. It's documented in all code supporting the original and the mini stream decks I've read. Trying to send the image in two halves gives nothing on my mini.

**Comment by cmsj** [comment]
@latenitefilms have you been able to try this on the full size Deck? I'm travelling atm and not able to test it, but if not then I'll take a look when I get home.

I'm curious to see if the image writing code works!

**Comment by latenitefilms** [comment]
@cmsj - Will test tonight and let you know.

Also:

Screen Shot 2019-05-27 at 8 01 26 am

**Comment by octplane** [comment]
The code seems to work ok on my mini now, if not paying too much attention to its speed. Last commit 6058e0176c6458d4d5f6d44e487a9b7c0c4bd636 should be parametrized because I expect it to break display on the regular Stream Deck. I had not time to do that properly yet...

**Comment by octplane** [comment]
After more Wiresharking, I finally find out that the continuation bool was not toggled in the right order, creating some weird issues in the way image were rendering, and esp. the performance hit.

It now works well and is ready for review. I have added a `buttonCount` method to the lua interface.

I'm joining a very simple `init.lua` that showcases displaying emojis (cheapest form of icons available) in the Deck. On my mini, although there is _not_ cache involved in that code, the performance is indiscernible from the original StreamDeck app's.

It should work also out of the box on a regular streamdeck, unless some of the constants for rotating/scaling... are wrong for this setup.

**Comment by octplane** [comment]
```lua
local c = require("hs.canvas")

function doIconSized(text, size)
  a = c.new {x = 0, y = 0, w = 80, h = 80}

  a[1] = {
    frame = {h = 80, w = 80, x = 0, y = 0},
    text = hs.styledtext.new(
      text,
      {
        font = {name = ".AppleSystemUIFont", size = size},
        paragraphStyle = {alignment = "center"}
      }
    ),
    type = "text"
  }
  return a:imageFromCanvas()
end

function doIcon(text)
  return doIconSized(text, 70)
end

function n(txt, content, action)
  local n = {}
  n.text = txt
  n.content = content
  return n
end

start = 0
treeNode = {
  "0️⃣",
  "1️⃣",
  "2️⃣",
  "3️⃣",
  "4️⃣",
  "5️⃣",
  "6️⃣",
  "🚨",
  "🏡",
  "🐶",
  "🌊",
  "✅",
  "🤡",
  "🚌",
  "🔥",
  "🏟"
}

function deckClicked(deck, i, pressed)
  if pressed then
    deck:setButtonImage(i, doIconSized(treeNode[i], 50))
  else
    deck:setButtonImage(i, doIcon(treeNode[i]))
    start = start + 1
    draw(deck, start)
  end
end

function draw(deck, from)
  for i = 1, deck:buttonCount() do
    local ix = (i + from) % #treeNode + 1
    print(ix)
    deck:setButtonImage(i, doIcon(treeNode[ix]))
  end
end

function cb(connected, deck)
  deck:reset()
  deck:buttonCallback(deckClicked)
  draw(deck, 0)
end

hs.streamdeck.init(cb)

```

**Comment by latenitefilms** [comment]
Awesome! Will test with the original Stream Deck tonight. Will also try get my hands on the mini and XL to test. It would be great if it also worked with the new iOS app too.

**Comment by latenitefilms** [comment]
Something is a bit funky @octplane. 

This is what your test code looks like on the original Stream Deck:

IMG_4079

**Comment by octplane** [comment]
that’s weird indeed. As if something was breaking mid course.

**Comment by cmsj** [comment]
Yikes, there’s also a StreamDeck XL now. I would hope it’s not a third completely different device, but I’m unlikely to buy one to find out.

**Comment by latenitefilms** [comment]
Yeah, if they keep adding more and more hardware options, it's going to get harder and harder to support. In CommandPost I'm considering dropping `hs.streamdeck` and moving towards using the official API to keep things easier.

However, it's worth noting that both [StreamDeckSharp](https://github.com/OpenMacroBoard/StreamDeckSharp) and [node-elgato-stream-deck](https://github.com/Lange/node-elgato-stream-deck) has added support for Stream Deck XL, so you might be able to go back to that code and see what's changed?

**Comment by octplane** [comment]
Yes, having at that again tonight...

**Comment by octplane** [comment]
So..., this is rather complicated ;)

I've read the others implementation code and one big change I did was to switch from Event based to Report based for the upstream SteamDeck->hs communication.

Our original implementation used Events and this was working ok, but was introducing crappy offsets to compute the correct button that was pressed.

The Report based implementation (used by the node.js and c# impl.) is much simpler as there is a simple offset to get the list of keys and their state (not exactly true as the original Deck seems to add some custom `transformKeyIndex` code). One big change is that we now get the state of all buttons at once, for every event.

This is cool, but we will have to store the state of all the buttons in our object and generate synthetics events for the lua client side.

So the good news is that you should be able to see some nice debug when you press buttons on your deck and that should be the right ones. cc @latenitefilms.

For the rest, I've started working on cleaning the code and trying to take in account mini, standard and XL variants everywhere, but haven't tested any code at all.

One overall interesting thing is that the implementation of the newer Decks is getting simpler as new versions arrive. So most of the complicated code is getting simpler (although I'm not 100% convinced this is the case for the image manipulation).

Short roadmap:

- [ ] make reset work (it still doesn't work although code looks good)
- [ ] dig on multi message communication structure (I had to hack stuff to make this work on my Deck but the other implementations seem simpler)
- [ ] work on bitmap transfer
- [ ] refactor key event / interface

**Comment by octplane** [comment]
At this point, it works again correctly on my Mini. The XL interface is missing (as payloads don't have the same structure as the 2 others.

Not sure about what its gives on the original.

Also `reset` do not work correctly on my Deck, shows a truncated logo. not sure why...

**Comment by latenitefilms** [comment]
Sorry for the lack of contact - I just tried this code, and can confirm the button assignment works correctly with the original Stream Deck on Mojave.

I did get one error however when initialising the extension:

```lua
2019-10-15 10:58:59: Welcome to the Hammerspoon Console!
You can run any Lua code in here.

2019-10-15 10:58:59: -- Lazy extension loading enabled
2019-10-15 10:59:00: -- Loading ~/.hammerspoon/init.lua
2019-10-15 10:59:00: init.lua loaded
2019-10-15 10:59:00: -- Done.

> hs.streamdeck.init(function(connected, obj)
    print(string.format("connected: %s", connected))
    print(string.format("obj: %s", obj))
end)
sd = hs.streamdeck.getDevice(1):buttonCallback(function(obj, buttonID, pressed)
    print(string.format("obj: %s", obj))
    print(string.format("buttonID: %s", buttonID))
    print(string.format("pressed: %s", pressed))
end)
2019-10-15 10:59:06: -- Loading extension: streamdeck
NSRangeException: *** -[__NSArrayM objectAtIndexedSubscript:]: index 0 beyond bounds for empty array
(
	0   CoreFoundation                      0x00007fff2bc5f2fd __exceptionPreprocess + 256
	1   libobjc.A.dylib                     0x00007fff56359a17 objc_exception_throw + 48
	2   CoreFoundation                      0x00007fff2bc9e2f4 -[CFPrefsConfigurationFileSource initWithConfigurationPropertyList:containingPreferences:] + 0
	3   CoreFoundation                      0x00007fff2bbef13b -[__NSArrayM objectAtIndexedSubscript:] + 151
	4   internal.so                         0x000000010ac8e728 streamdeck_getDevice + 1064
	5   LuaSkin                             0x00000001018bab27 luaD_precall + 5383
	6   LuaSkin                             0x000000010189faa8 luaV_execute + 146088
	7   LuaSkin                             0x00000001018bed42 luaD_call + 290
	8   LuaSkin                             0x00000001018bf164 luaD_callnoyield + 212
	9   LuaSkin                             0x000000010185485c f_call + 428
	10  LuaSkin                             0x00000001017872ad luai_objcttry + 445
	11  LuaSkin                             0x00000001018b35a5 luaD_rawrunprotected + 1173
	12  LuaSkin                             0x00000001018c4f13 luaD_pcall + 1107
	13  LuaSkin                             0x00000001018530a0 lua_pcallk + 5168
	14  LuaSkin                             0x00000001018b044f luaB_xpcall + 207
	15  LuaSkin                             0x00000001018bab27 luaD_precall + 5383
	16  LuaSkin                             0x000000010189faa8 luaV_execute + 146088
	17  LuaSkin                             0x00000001018bed42 luaD_call + 290
	18  LuaSkin                             0x00000001018bf164 luaD_callnoyield + 212
	19  LuaSkin                             0x000000010185485c f_call + 428
	20  LuaSkin                             0x00000001017872ad luai_objcttry + 445
	21  LuaSkin                             0x00000001018b35a5 luaD_rawrunprotected + 1173
	22  LuaSkin                             0x00000001018c4f13 luaD_pcall + 1107
	23  LuaSkin                             0x00000001018530a0 lua_pcallk + 5168
	24  LuaSkin                             0x0000000101750db7 -[LuaSkin protectedCallAndTraceback:nresults:] + 727
	25  Hammerspoon                         0x0000000100023d26 MJLuaRunString + 1222
	26  Hammerspoon                         0x0000000100017fcb -[MJConsoleWindowController run:] + 363
	27  Hammerspoon                         0x0000000100018478 -[MJConsoleWindowController tryMessage:] + 1032
	28  AppKit                              0x00007fff29491644 -[NSApplication(NSResponder) sendAction:to:from:] + 312
	29  AppKit                              0x00007fff294fb992 -[NSControl sendAction:to:] + 86
	30  AppKit                              0x00007fff2948d68f -[NSTextField textDidEndEditing:] + 924
	31  Hammerspoon                         0x000000010001cb3a -[HSGrowingTextField textDidEndEditing:] + 602
	32  CoreFoundation                      0x00007fff2bc0bb96 __CFNOTIFICATIONCENTER_IS_CALLING_OUT_TO_AN_OBSERVER__ + 12
	33  CoreFoundation                      0x00007fff2bc0bb10 ___CFXRegistrationPost_block_invoke + 63
	34  CoreFoundation                      0x00007fff2bc0ba7a _CFXRegistrationPost + 404
	35  CoreFoundation                      0x00007fff2bc13f28 ___CFXNotificationPost_block_invoke + 87
	36  CoreFoundation                      0x00007fff2bb7c8b4 -[_CFXNotificationRegistrar find:object:observer:enumerator:] + 1642
	37  CoreFoundation                      0x00007fff2bb7bc67 _CFXNotificationPost + 732
	38  Foundation                          0x00007fff2de01aab -[NSNotificationCenter postNotificationName:object:userInfo:] + 66
	39  AppKit                              0x00007fff2959defb -[NSTextView(NSPrivate) _giveUpFirstResponder:] + 415
	40  AppKit                              0x00007fff294caca6 -[NSTextView doCommandBySelector:] + 194
	41  AppKit                              0x00007fff294cabba -[NSTextInputContext(NSInputContext_WithCompletion) doCommandBySelector:completionHandler:] + 228
	42  AppKit                              0x00007fff294bfe1a -[NSKeyBindingManager(NSKeyBindingManager_MultiClients) interpretEventAsCommand:forClient:] + 2972
	43  AppKit                              0x00007fff29b8bfdf __84-[NSTextInputContext _handleEvent:options:allowingSyntheticEvent:completionHandler:]_block_invoke_5 + 341
	44  AppKit                              0x00007fff29b8be75 __84-[NSTextInputContext _handleEvent:options:allowingSyntheticEvent:completionHandler:]_block_invoke_3.784 + 74
	45  AppKit                              0x00007fff294c6f23 -[NSTextInputContext tryHandleEvent_HasMarkedText_withDispatchCondition:dispatchWork:continuation:] + 87
	46  AppKit                              0x00007fff29b8bdfb __84-[NSTextInputContext _handleEvent:options:allowingSyntheticEvent:completionHandler:]_block_invoke.781 + 237
	47  HIToolbox                           0x00007fff2ae6fefb __TSMProcessRawKeyEventWithOptionsAndCompletionHandler_block_invoke_5 + 70
	48  HIToolbox                           0x00007fff2ae6edb9 ___ZL23DispatchEventToHandlersP14EventTargetRecP14OpaqueEventRefP14HandlerCallRec_block_invoke + 110
	49  AppKit                              0x00007fff29b8670e __55-[NSTextInputContext handleTSMEvent:completionHandler:]_block_invoke.265 + 575
	50  AppKit                              0x00007fff294c1512 __55-[NSTextInputContext handleTSMEvent:completionHandler:]_block_invoke_2 + 74
	51  AppKit                              0x00007fff294c149a -[NSTextInputContext tryHandleTSMEvent_HasMarkedText_withDispatchCondition:dispatchWork:continuation:] + 87
	52  AppKit                              0x00007fff294c0c92 -[NSTextInputContext handleTSMEvent:completionHandler:] + 1749
	53  AppKit                              0x00007fff294c0545 _NSTSMEventHandler + 306
	54  HIToolbox                           0x00007fff2ae0522e _ZL23DispatchEventToHandlersP14EventTargetRecP14OpaqueEventRefP14HandlerCallRec + 1422
	55  HIToolbox                           0x00007fff2ae045df _ZL30SendEventToEventTargetInternalP14OpaqueEventRefP20OpaqueEventTargetRefP14HandlerCallRec + 371
	56  HIToolbox                           0x00007fff2ae04465 SendEventToEventTargetWithOptions + 45
	57  HIToolbox                           0x00007fff2ae6bf8f SendTSMEvent_WithCompletionHandler + 383
	58  HIToolbox                           0x00007fff2ae6c3fa __SendUnicodeTextAEToUnicodeDoc_WithCompletionHandler_block_invoke + 387
	59  HIToolbox                           0x00007fff2ae6c268 __SendFilterTextEvent_WithCompletionHandler_block_invoke + 221
	60  HIToolbox                           0x00007fff2ae6bfde SendTSMEvent_WithCompletionHandler + 462
	61  HIToolbox                           0x00007fff2ae6bde3 SendFilterTextEvent_WithCompletionHandler + 225
	62  HIToolbox                           0x00007fff2ae6baa4 SendUnicodeTextAEToUnicodeDoc_WithCompletionHandler + 280
	63  HIToolbox                           0x00007fff2ae6b84e __utDeliverTSMEvent_WithCompletionHandler_block_invoke_2 + 283
	64  HIToolbox                           0x00007fff2ae6b6ad __utDeliverTSMEvent_WithCompletionHandler_block_invoke + 355
	65  HIToolbox                           0x00007fff2ae6b4cb TSMKeyEvent_WithCompletionHandler + 598
	66  HIToolbox                           0x00007fff2ae6b25a __TSMProcessRawKeyEventWithOptionsAndCompletionHandler_block_invoke_4 + 250
	67  HIToolbox                           0x00007fff2ae6b089 __TSMProcessRawKeyEventWithOptionsAndCompletionHandler_block_invoke_3 + 257
	68  HIToolbox                           0x00007fff2ae6adce __TSMProcessRawKeyEventWithOptionsAndCompletionHandler_block_invoke_2 + 282
	69  HIToolbox                           0x00007fff2ae6ab32 __TSMProcessRawKeyEventWithOptionsAndCompletionHandler_block_invoke + 274
	70  HIToolbox                           0x00007fff2ae6a127 TSMProcessRawKeyEventWithOptionsAndCompletionHandler + 3398
	71  AppKit                              0x00007fff29b8bcd9 __84-[NSTextInputContext _handleEvent:options:allowingSyntheticEvent:completionHandler:]_block_invoke_3.779 + 110
	72  AppKit                              0x00007fff29b8b2d6 __204-[NSTextInputContext tryTSMProcessRawKeyEvent_orSubstitution:dispatchCondition:setupForDispatch:furtherCondition:doubleSpaceSubstitutionCondition:doubleSpaceSubstitutionWork:dispatchTSMWork:continuation:]_block_invoke.734 + 115
	73  AppKit                              0x00007fff29b8b1c7 -[NSTextInputContext tryTSMProcessRawKeyEvent_orSubstitution:dispatchCondition:setupForDispatch:furtherCondition:doubleSpaceSubstitutionCondition:doubleSpaceSubstitutionWork:dispatchTSMWork:continuation:] + 245
	74  AppKit                              0x00007fff29b8b892 -[NSTextInputContext _handleEvent:options:allowingSyntheticEvent:completionHandler:] + 1286
	75  AppKit                              0x00007fff29b8b086 -[NSTextInputContext _handleEvent:allowingSyntheticEvent:] + 107
	76  AppKit                              0x00007fff294bf004 -[NSView interpretKeyEvents:] + 209
	77  AppKit                              0x00007fff294bee2d -[NSTextView keyDown:] + 726
	78  AppKit                              0x00007fff2930c367 -[NSWindow(NSEventRouting) _reallySendEvent:isDelayedEvent:] + 6840
	79  AppKit                              0x00007fff2930a667 -[NSWindow(NSEventRouting) sendEvent:] + 478
	80  AppKit                              0x00007fff291aa8a9 -[NSApplication(NSEvent) sendEvent:] + 2953
	81  AppKit                              0x00007fff291985e0 -[NSApplication run] + 755
	82  AppKit                              0x00007fff29187ae8 NSApplicationMain + 777
	83  Hammerspoon                         0x0000000100011774 main + 52
	84  libdyld.dylib                       0x00007fff57b273d5 start + 1
	85  ???                                 0x0000000000000005 0x0 + 5
)
2019-10-15 10:59:06: connected: true
2019-10-15 10:59:06: obj: hs.streamdeck: Elgato Stream Deck, serial: AL29G1A09928 (0x604000665af8)
2019-10-15 10:59:20: ********
2019-10-15 10:59:20: 10:59:20 ERROR:   LuaSkin: hs.streamdeck received a button input, but no callback has been set. See hs.streamdeck:buttonCallback()
2019-10-15 10:59:20: ********
2019-10-15 10:59:20: ********
2019-10-15 10:59:20: 10:59:20 ERROR:   LuaSkin: hs.streamdeck received a button input, but no callback has been set. See hs.streamdeck:buttonCallback()
2019-10-15 10:59:20: ********

> sd = hs.streamdeck.getDevice(1):buttonCallback(function(obj, buttonID, pressed)
    print(string.format("obj: %s", obj))
    print(string.format("buttonID: %s", buttonID))
    print(string.format("pressed: %s", pressed))
end)

2019-10-15 10:59:28: obj: hs.streamdeck: Elgato Stream Deck, serial: AL29G1A09928 (0x6040002727f8)
2019-10-15 10:59:28: buttonID: 1
2019-10-15 10:59:28: pressed: true
2019-10-15 10:59:29: obj: hs.streamdeck: Elgato Stream Deck, serial: AL29G1A09928 (0x6040002ac778)
2019-10-15 10:59:29: buttonID: 1
2019-10-15 10:59:29: pressed: false
2019-10-15 10:59:30: obj: hs.streamdeck: Elgato Stream Deck, serial: AL29G1A09928 (0x6040002c50f8)
2019-10-15 10:59:30: buttonID: 2
2019-10-15 10:59:30: pressed: true
2019-10-15 10:59:30: obj: hs.streamdeck: Elgato Stream Deck, serial: AL29G1A09928 (0x6040006db738)
2019-10-15 10:59:30: buttonID: 2
2019-10-15 10:59:30: pressed: false
2019-10-15 10:59:30: obj: hs.streamdeck: Elgato Stream Deck, serial: AL29G1A09928 (0x6040006ddbb8)
2019-10-15 10:59:30: buttonID: 3
2019-10-15 10:59:30: pressed: true
2019-10-15 10:59:30: obj: hs.streamdeck: Elgato Stream Deck, serial: AL29G1A09928 (0x6040006dfb38)
2019-10-15 10:59:30: buttonID: 3
2019-10-15 10:59:30: pressed: false
2019-10-15 10:59:30: obj: hs.streamdeck: Elgato Stream Deck, serial: AL29G1A09928 (0x6040006e1a78)
2019-10-15 10:59:30: buttonID: 4
2019-10-15 10:59:30: pressed: true
2019-10-15 10:59:31: obj: hs.streamdeck: Elgato Stream Deck, serial: AL29G1A09928 (0x6040006e3f38)
2019-10-15 10:59:31: buttonID: 4
2019-10-15 10:59:31: pressed: false
2019-10-15 10:59:31: obj: hs.streamdeck: Elgato Stream Deck, serial: AL29G1A09928 (0x6040006e5df8)
2019-10-15 10:59:31: buttonID: 5
2019-10-15 10:59:31: pressed: true
2019-10-15 10:59:31: obj: hs.streamdeck: Elgato Stream Deck, serial: AL29G1A09928 (0x6040006a8878)
2019-10-15 10:59:31: buttonID: 5
2019-10-15 10:59:31: pressed: false
2019-10-15 10:59:31: obj: hs.streamdeck: Elgato Stream Deck, serial: AL29G1A09928 (0x6040006e6f38)
2019-10-15 10:59:31: buttonID: 6
2019-10-15 10:59:31: pressed: true
2019-10-15 10:59:32: obj: hs.streamdeck: Elgato Stream Deck, serial: AL29G1A09928 (0x6040006e8eb8)
2019-10-15 10:59:32: buttonID: 6
2019-10-15 10:59:32: pressed: false
2019-10-15 10:59:32: obj: hs.streamdeck: Elgato Stream Deck, serial: AL29G1A09928 (0x604000394bf8)
2019-10-15 10:59:32: buttonID: 7
2019-10-15 10:59:32: pressed: true
2019-10-15 10:59:32: obj: hs.streamdeck: Elgato Stream Deck, serial: AL29G1A09928 (0x6040003d90b8)
2019-10-15 10:59:32: buttonID: 7
2019-10-15 10:59:32: pressed: false
2019-10-15 10:59:32: obj: hs.streamdeck: Elgato Stream Deck, serial: AL29G1A09928 (0x6040003bc078)
2019-10-15 10:59:32: buttonID: 8
2019-10-15 10:59:32: pressed: true
2019-10-15 10:59:32: obj: hs.streamdeck: Elgato Stream Deck, serial: AL29G1A09928 (0x6040003e6778)
2019-10-15 10:59:32: buttonID: 8
2019-10-15 10:59:32: pressed: false
2019-10-15 10:59:33: obj: hs.streamdeck: Elgato Stream Deck, serial: AL29G1A09928 (0x6040003d4038)
2019-10-15 10:59:33: buttonID: 9
2019-10-15 10:59:33: pressed: true
2019-10-15 10:59:33: obj: hs.streamdeck: Elgato Stream Deck, serial: AL29G1A09928 (0x6040006ebe38)
2019-10-15 10:59:33: buttonID: 9
2019-10-15 10:59:33: pressed: false
2019-10-15 10:59:33: obj: hs.streamdeck: Elgato Stream Deck, serial: AL29G1A09928 (0x6040006ee178)
2019-10-15 10:59:33: buttonID: 10
2019-10-15 10:59:33: pressed: true
2019-10-15 10:59:33: obj: hs.streamdeck: Elgato Stream Deck, serial: AL29G1A09928 (0x6040006f1478)
2019-10-15 10:59:33: buttonID: 10
2019-10-15 10:59:33: pressed: false
2019-10-15 10:59:34: obj: hs.streamdeck: Elgato Stream Deck, serial: AL29G1A09928 (0x6040006f38b8)
2019-10-15 10:59:34: buttonID: 11
2019-10-15 10:59:34: pressed: true
2019-10-15 10:59:34: obj: hs.streamdeck: Elgato Stream Deck, serial: AL29G1A09928 (0x6040006f5678)
2019-10-15 10:59:34: buttonID: 11
2019-10-15 10:59:34: pressed: false
2019-10-15 10:59:34: obj: hs.streamdeck: Elgato Stream Deck, serial: AL29G1A09928 (0x6040006f79b8)
2019-10-15 10:59:34: buttonID: 12
2019-10-15 10:59:34: pressed: true
2019-10-15 10:59:34: obj: hs.streamdeck: Elgato Stream Deck, serial: AL29G1A09928 (0x6040006f9778)
2019-10-15 10:59:34: buttonID: 12
2019-10-15 10:59:34: pressed: false
2019-10-15 10:59:35: obj: hs.streamdeck: Elgato Stream Deck, serial: AL29G1A09928 (0x6040006fba78)
2019-10-15 10:59:35: buttonID: 13
2019-10-15 10:59:35: pressed: true
2019-10-15 10:59:35: obj: hs.streamdeck: Elgato Stream Deck, serial: AL29G1A09928 (0x6040006fc1f8)
2019-10-15 10:59:35: buttonID: 13
2019-10-15 10:59:35: pressed: false
2019-10-15 10:59:35: obj: hs.streamdeck: Elgato Stream Deck, serial: AL29G1A09928 (0x6040006fdfb8)
2019-10-15 10:59:35: buttonID: 14
2019-10-15 10:59:35: pressed: true
2019-10-15 10:59:36: obj: hs.streamdeck: Elgato Stream Deck, serial: AL29G1A09928 (0x6040006ffdf8)
2019-10-15 10:59:36: buttonID: 14
2019-10-15 10:59:36: pressed: false
2019-10-15 10:59:36: obj: hs.streamdeck: Elgato Stream Deck, serial: AL29G1A09928 (0x604000701b78)
2019-10-15 10:59:36: buttonID: 15
2019-10-15 10:59:36: pressed: true
2019-10-15 10:59:36: obj: hs.streamdeck: Elgato Stream Deck, serial: AL29G1A09928 (0x604000703938)
2019-10-15 10:59:36: buttonID: 15
2019-10-15 10:59:36: pressed: false
```

**Comment by latenitefilms** [comment]
Actually upon further testing, this makes buttons 3, 4 and 5 blue - so there's something funky going on:

```lua
> sd:setButtonColor(1, hs.drawing.color.blue)
hs.streamdeck: Elgato Stream Deck, serial: AL29G1A09928 (0x6040003fd2b8)

> sd:setButtonColor(2, hs.drawing.color.blue)
hs.streamdeck: Elgato Stream Deck, serial: AL29G1A09928 (0x6040003fd2b8)

> sd:setButtonColor(3, hs.drawing.color.blue)
hs.streamdeck: Elgato Stream Deck, serial: AL29G1A09928 (0x6040003fd2b8)
```

**Comment by latenitefilms** [comment]
And this makes buttons 1 & 2 turn blue:

```lua
> sd:setButtonColor(0, hs.drawing.color.blue)
hs.streamdeck: Elgato Stream Deck, serial: AL29G1A09928 (0x6040003fd2b8)

> sd:setButtonColor(-1, hs.drawing.color.blue)
hs.streamdeck: Elgato Stream Deck, serial: AL29G1A09928 (0x6040003fd2b8)
```

**Comment by latenitefilms** [comment]
This might be helpful:

https://github.com/timothycrosley/streamdeck-ui/

**Comment by cmsj** [comment]
I just picked up a Stream Deck Mini, so I'm happy to take over getting this all into shape - I'll start by picking apart https://github.com/abcminiuser/python-elgato-streamdeck/tree/master/src/StreamDeck to see what the differences are with the Mini, and since they also support the XL, we could have some untested support for that (but I'm definitely not going to spend the $200 on an XL to find out if my code works there :) )

**Comment by octplane** [comment]
Thanks @cmsj , I confess my personal time has greatly vanished these past months and I reverted back to the use of the built-in app, even if it's sadly less useful than HS.

Happy to send you 🍺  over for the work, though, 🙇

**Comment by cmsj** [comment]
@octplane I've just pushed up https://github.com/cmsj/hammerspoon/tree/refactor-streamdeck which works for me on both the original deck and the mini. What I did was refactor HSStreamDeckDevice to be a base class, with subclasses for each of the variants, rather than have all the differences inline. The reason for that being that, in looking at https://github.com/Hammerspoon/hammerspoon/pull/2102 I realised that there are actually now 4 hardware variants to deal with. For now I've just got Original and Mini supported, because I don't have an Original v2 or an XL, and I haven't built in some of the additional changes you made.

Maybe you could help me understand some of the changes I'm missing, and we can get this finished/landed? I'd then be willing to carry untested support for the V2 and the XL, and we'll see if anyone ever tries them or not.

**Comment by cmsj** [comment]
Ok, I think I'm catching up here, but I'm still missing XL/v2 support and the HID Report support. I'm still not 100% sold on the HID Report stuff, because of the added complexity of storing all the button states. I'm wondering if the HID events are stable enough that we could just apply an offset to the numbers?

**Comment by cmsj** [comment]
Thanks very much @octplane - I really appreciate you starting this PR, even though I ended up rewriting it pretty significantly. You provided some very valuable reference code to work from :)

**Pull request closed without merging**